eng/azure-pipelines-codeql.yml

```diff
@@ -0,0 +1,102 @@
+trigger:
+  none
+
+schedules:
+  - cron: 0 12 * * 1
+    displayName: Weekly Monday CodeQL/Semmle run
+    branches:
+      include:
+      - dotnet/main
+      - objwriter/12.x
+    always: true
+
+variables:
+  - template: /eng/common-variables.yml
+  - name: Codeql.Enabled
+    value: True
+  - name: Codeql.Cadence
+    value: 0
+  - name: Codeql.TSAEnabled
+    value: True
+  - name: Codeql.BuildIdentifier
+    value: $(System.JobDisplayName)
+
+stages:
+- stage: Build
+  jobs:
+  - template: /eng/common/templates/jobs/jobs.yml
+    parameters:
+      jobs:
+
+      ############ LINUX BUILD ############
+      - job: Build_Linux
+        displayName: Linux
+        timeoutInMinutes: 480
+        variables:
+        - _BuildConfig: Release
+        strategy:
+          matrix:
+            x64:
+              imagename: mcr.microsoft.com/dotnet-buildtools/prereqs:centos-7-20220716123527-d0bc8ed
+              archflag: --arch x64
+              Devtoolset7Arg: /p:ForceDevtoolset7=true
+        pool:
+          ${{ if eq(variables['System.TeamProject'], 'public') }}:
+            vmImage: ubuntu-20.04
+          ${{ if eq(variables['System.TeamProject'], 'internal') }}:
+            name: NetCore1ESPool-Internal
+            demands: ImageOverride -equals Build.Ubuntu.1804.Amd64
+        container:
+          image: $(imagename)
+        steps:
+        - bash: |
+            set -ex
+            git clean -ffdx
+            git reset --hard HEAD
+          displayName: 'Clean up working directory'
+
+        - task: CodeQL3000Init@0
+          displayName: Initialize CodeQL (manually-injected)
+
+        - bash: |
+            ./build.sh --ci --restore --build $(archflag) --configuration $(_BuildConfig) $(_InternalBuildArgs) $(Devtoolset7Arg)
```

Comment:
Is this a common cron to build the repo or does it run something else after that? Could you explain to me how codeql does its checks?

Reply:
Typically CodeQL just wraps around the build that is compiled to produce the artifacts needed to support a build/service. CodeQL3000Init@0 and CodeQL3000Finalize@0 are just manual invocations of the typical CodeQL that is automatically injected in the default branch pipeline builds. If you check the `llvm-project-official` pipeline, you may spot the CodeQL auto injections.

However, because this PR is opting to enable CodeQL in a separate pipeline in order to not impact the official pipeline too much (especially because builds with CodeQL enabled take significantly longer to finish because CodeQL scans all detectable languages) we need to wrap the build process manually with the tasks.

I dont know all the specifics of how codeql actively functions, but I believe it scans all code that gets compiled during the tasks that its wrapped around.

Reply:
Thanks a lot for the explanation. Where do we check the list of warnings/suggestions this pipeline has found?

Reply:
In the tsaoptions.json, there is a notification alias that is pinged with what I believe is a summary of all of the warnings/issues found during the scan. Right now I just put the infrastructure folks by default, but if you or your team want to have the pings instead we can change it!

Besides getting the notification, those with permission to view the internal pipelines can take at the pipeline (dotnet-llvm-project-codeql), click on a build, choose any job of the build stage, and scroll to the bottom of the FInalize CodeQL step to see the upload details of the CodeQL scan, of which there should be links to the `LGTM Analysis Status`